lib/ParseSIL/ParseSIL.cpp

```diff
@@ -3117,6 +3117,29 @@ bool SILParser::parseSILInstruction(SILBuilder &B) {
   }
   // SWIFT_ENABLE_TENSORFLOW
   case SILInstructionKind::GraphOperationInst: {
+    bool noClustering = false;
+    if (P.consumeIf(tok::l_square)) {
+      Identifier ident;
+      SourceLoc identLoc;
+      if (parseSILIdentifier(ident, identLoc,
+                             diag::expected_in_attribute_list)) {
+        P.diagnose(P.Tok, diag::expected_tok_in_sil_instr,
+                   "'no_clustering' attribute");
+        return true;
+      }
+      StringRef attr = ident.str();
+
+      if (attr == "no_clustering")
```

Comment:
Consider using `parseSpecificIdentifier`? Example usage:
```cpp
 if (P.parseSpecificIdentifier(
        "source", diag::sil_attr_differentiable_expected_keyword, "source"))
  return true;
```

Reply:
Or if that doesn't work,  consider `if (ident == "no_clustering")` instead of declaring local variable `attr`?
(Not sure if `Identifier` supports `==`, if not try `if (ident.str() == "no_clustering")`.)

Reply:
I used Richard's suggestion below to simplify the code here quite a bit. Also `ident.str() == "no_clustering"` works. PTAL. Thanks!

Reply:
Looks good!